apps/sms/services/js/conversation/conversation_service.js

```diff
@@ -1,47 +1,184 @@
+ * messages that user hasn't seen yet.
+ * @property {string} lastMessageType Type of the last message in the
+ * conversation. Can be either 'sms' or 'mms'.
+ * @property {Draft} draft If conversation is draft or has draft, then this
+ * field contains appropriate draft information, otherwise it's null.
+ */
+
+[
+  ['bridge', '/lib/bridge/bridge.js'],
+  ['streamClient', '/lib/bridge/plugins/stream/client.js'],
+  ['BridgeServiceMixin', '/services/js/bridge_service_mixin.js'],
+  ['Drafts', '/services/js/drafts.js']
+].forEach(([dependencyName, dependencyPath]) => {
+  if (!(dependencyName in self)) {
+    importScripts(dependencyPath);
+  }
+});
 
 (function(exports) {
-  const SERVICE_NAME = 'conversation-service';
+  const debug = 0 ?
+    (arg1, ...args) => console.log(`[ConversationService] ${arg1}`, ...args):
+    () => {};
 
-  const STREAMS = Object.freeze(
-    ['getAllConversations', 'getMessagesForConversation']
-  );
+  const mark = 0 ?
+    (markName) =>
+      performance.mark(`[ConversationService] ${markName}@sms.gaiamobile.org`):
+    () => {};
+
+  const priv = {
+    mobileMessageClients: Symbol('mobileMessageClients'),
+
+    getMobileMessageClient: Symbol('getMobileMessageClient')
+  };
+
+  const SERVICE_CONTRACT = Object.freeze({
+    name: 'conversation-service',
+
+    streams: Object.freeze(
+      ['getAllConversations', 'getMessagesForConversation']
+    ),
+
+    methods: Object.freeze([
+      'deleteConversations', 'deleteMessages', 'markConversationsAs',
+      'getConversationSummary', 'getMessage', 'findConversationFromAddress'
+    ]),
 
-  const METHODS = Object.freeze([
-    'deleteConversations', 'deleteMessages', 'markConversationsAs',
-    'getConversationSummary', 'getMessage', 'findConversationFromAddress'
-  ]);
+    events: Object.freeze(['message-change'])
+  });
 
-  const EVENTS = Object.freeze([
-    'message-change'
-  ]);
+  function draftToConversationSummary(draft) {
+    var body = Array.isArray(draft.content) ?
+      draft.content.find((content) => typeof content === 'string') : '';
+
+    return {
+      id: +draft.id,
+      participants: Array.isArray(draft.recipients) ? draft.recipients : [],
+      body: body,
+      timestamp: new Date(draft.timestamp),
+      status: { hasUnread: false, hasNewError: false },
+      lastMessageType: draft.type || 'sms',
+      draft: new Draft(draft)
+    };
+  }
+
+  function threadToConversationSummary(thread) {
+    return {
+      id: thread.id,
+      participants: thread.participants,
+      body: thread.body,
+      timestamp: thread.timestamp,
+      status: { hasUnread: thread.unreadCount > 0, hasNewError: false },
+      lastMessageType: thread.lastMessageType,
+      draft: Drafts.byThreadId(thread.id)
+    };
+  }
 
   var ConversationService = {
     /**
+     * List of mobileMessageClients mapped to specific application id.
+     * @type {Map.<string, Client>}
+     */
+    [priv.mobileMessageClients]: null,
```

Comment:
Well, I could potentially use ```WeakMap.<ConversationClient, MobileMessageClient>``` here with the snippet like:
```
/**
 * @type {WeakMap.<Client, Client>}
 */
[priv.mobileMessageClients]: null,

getAllConversations(serviceStream, appInstanceId) {
  var mobileMessageClient = this[priv.getMobileMessageClient](serviceStream.clientId, appInstanceId);
  ....
},

[priv.getMobileMessageClient](clientId, appInstanceId) {
  var conversationClient = this.service.getClient(clientId);

  var mobileMessageClient = this[priv.mobileMessageClients].get(
    conversationClient
  );

  if (!mobileMessageClient) {
    mobileMessageClient = bridge.client({
      service: 'mozMobileMessageShim',
      endpoint: new BroadcastChannel('shim-channel-' + appInstanceId)
    }).plugin(streamClient);

    this[priv.mobileMessageClients].set(conversationClient, mobileMessageClient);

    debug(
      'Create MobileMessageClient for app instance %s', appInstanceId
    );
  }

  return mobileMessageClient;
}
```

That way we won't have dead bridge client references, but it required changes in bridge mixin to expose ```getClient``` method and I'm not 100% sure how long we'll live with this workaround, so I turned on ```lazy-mode``` and used the simplest approach :)